### PR TITLE
Fix negation function name collision.

### LINKED ISSUE #11631: [Codegen] Code compiled via legacy codegen may be deployed successfully, but not via Sol->Yul (reverts)
```
contract C {
  constructor(int24 i0) {
    -(i0);
    unchecked {
      (-(type(int24).min));
    }
  }
}
// ====
// compileViaYul: also
// ----
// constructor(): 1 ->
```

The test above passes via legacy codegen but fails via Yul. I would expect the latter to also be successful.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/11631

### CODE REVIEW

**Comment by ekpyron** [comment]
~~Still needs at least a test case.~~ Added test.

**Comment by ekpyron** [comment]
Didn't we want to put some mechanism into place to prevent this kind of thing :-)?

File: Changelog.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ Bugfixes:
  * Type Checker: Fix internal error and prevent static calls to unimplemented modifiers.
  * Yul Code Generator: Fix internal compiler error when using a long literal with bitwise negation.
  * Yul Code Generator: Fix source location references for calls to builtin functions.
+ * Yul Code Generator: Use the correct unary negation function, if both checked and unchecked unary negation of the same type occurs.
```
Comment (chriseth) [inline Changelog.md]: No changelog, it has not been released yet. The other changelog entries are about generating code from yul.

**Comment by chriseth** [comment]
I'm happy for any mechanism that prevents this. We have to take the name of the current function into account and something like equality comparison on all the arguments.

File: Changelog.md (hunk 1 above)
Comment (ekpyron) [reply Changelog.md]: Removed.

**Comment by ekpyron** [comment]
> I'm happy for any mechanism that prevents this. We have to take the name of the current function into account and something like equality comparison on all the arguments.

Yeah, the reason why we don't have one yet, is probably that it's rather tricky to come up with a really nice and automatic one... the C++ Reflection TS would help and it would even be enough if any single compiler had a released version of it - one failing CI test run would already be enough - but it seems like both gcc and clang still keep their implementation of that on separate branches.
The only ugly-but-half-decent other option I see would be to define a macro to be used to defining these functions...
So instead of
```
string YulUtilFunctions::negateNumberWrappingFunction(Type const& _type)
{
...
}
```
we'd have to use something like
```
DECLARE_YUL_UTIL_FUNCTION(string, negateNumberWrappingFunction, Type const& _type)
{
...
}
```
Or we change the calling convention, s.t. ``m_utils.negateNumberWrappingFunction(type)`` has to go through a macro, but that's probably even more annoying, since autocompletion will probably be gone - albeit that way it'd be easier to guarantee that it's always done compared to us at least having to make sure that all these functions are in fact declared using the macro...

If we say, that's ugly, but better than having to be afraid of these bugs, I can write something like that...

**Comment by ekpyron** [comment]
Or alternatively: I'm not actually sure if it's actually worth it to avoid regenerating these functions on each call... if we'd be fine with that we could also simply have ``m_functionCollector.createFunction`` run the generator each time and, if we have already stored a function of the same name, just use string comparison to verify that it's the same one I guess. Might be worth checking, if that has any noticable performance impact (or maybe we did that and that's why we have the caching in the first place?).

**Comment by ekpyron** [comment]
Oh interesting, this seems to be actually happen in ``prbmath_signed`` and changes its gas costs!

**Comment by chriseth** [comment]
> Or alternatively: I'm not actually sure if it's actually worth it to avoid regenerating these functions on each call... if we'd be fine with that we could also simply have `m_functionCollector.createFunction` run the generator each time and, if we have already stored a function of the same name, just use string comparison to verify that it's the same one I guess. Might be worth checking, if that has any noticable performance impact (or maybe we did that and that's why we have the caching in the first place?).

That sounds like a good approach, but how do we handle functions again generated from that function?

**Review by chriseth — APPROVED** [review]

**Comment by ekpyron** [comment]
> > Or alternatively: I'm not actually sure if it's actually worth it to avoid regenerating these functions on each call... if we'd be fine with that we could also simply have `m_functionCollector.createFunction` run the generator each time and, if we have already stored a function of the same name, just use string comparison to verify that it's the same one I guess. Might be worth checking, if that has any noticable performance impact (or maybe we did that and that's why we have the caching in the first place?).
> 
> That sounds like a good approach, but how do we handle functions again generated from that function?

Shouldn't something like this: https://github.com/ethereum/solidity/pull/11635 be enough? It's based on ``develop`` before this being merged and it should catch this bug in the ``pbrmath`` test...

**Pull request merged**